healer/README.md

```diff
@@ -0,0 +1,124 @@
+# CSI-Addons Operation: Healer
+
+## Terminology
+
+| Term     | Definition                                                                            |
+| -------- | ------------------------------------------------------------------------------------- |
+| VolumeID | The identifier of the volume generated by the plugin.                                 |
+| CO       | Container Orchestration system that communicates with plugins using CSI service RPCs. |
+| SP       | Storage Provider, the vendor of a CSI plugin implementation.                          |
+| RPC      | [Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call).         |
+
+## Objective
+
+Define a standard that will enable storage providers (SP) to
+perform node level volume health check and healing operations.
+
+### Goals in MVP
+
+The new extension will define a procedure that
+
+* can be called for existing volumes
+* interacts with the Node-Plugin to check the health condition of the volume
+* makes it possible for the SP to heal the volumes if they are in abnormal
+  condition
+
+### Non-Goals in MVP
+
+* Implementation of healing logic is OPTIONAL and completely SP specific
+
+## Solution Overview
+
+This specification defines an interface along with the minimum operational and
+packaging recommendations for a storage provider (SP) to implement a
+health check and heal operations for volumes. The interface declares the
+RPCs that a plugin MUST expose.
+
+## RPC Interface
+
+* **Node Service**: The Node plugin MUST implement this RPC.
+
+```protobuf
+syntax = "proto3";
+package healer;
+
+import "github.com/container-storage-interface/spec/lib/go/csi/csi.proto";
+import "google/protobuf/descriptor.proto";
+
+option go_package = "github.com/csi-addons/spec/lib/go/healer";
+
+// HealerNode holds the RPC method for running heal operations on the
+// active (staged/published) volume.
+service HealerNode {
+  // NodeHealer is a procedure that gets called on the CSI NodePlugin.
+  rpc NodeHealer (NodeHealerRequest)
```

Comment:
Maybe call this `HealVolume`?

I would expect a way to check if a volume needs healing. The current `NodeHealer` call seems to do that, but _also_ is expected to do the healing?

It would probably be useful to have a check, and only perform healing if the check returns that it is needed. This then can provide better feedback and tracking of actions that were performed. It might even be useful to fallback to some other recovery/healing mechanism in case volume-healing failed (extension for `HealNode` which might do a reboot or inform medik8s or similar?)

Reply:
@nixpanic 
NodeHealer() is called on  CSI NodePlugin, so it's up to SP to implement/perform healing or not, right? if so why do we need a flag/check?

I'm probably missing something? Is this flag set as part of the request or response?
* If it is part of request then who will be setting it because we don't have any CRDs here.
* If it is part of the response, may be based on the response CSI addon might want to do a reboot or similar?


Could you please make it more clear?
Also cannot we detect healing failed from the message string which is part of the response?

Reply:
I would consider two procedures:

1. `GetHealth()` returning the state of the volume, possibly with a selection of recovery options (`VolumeHealing` or `NodeHealing` required)
2. `HealVolume()` doing the actual healing, if it can heal the volume at all
3. the CSI-Addons implementation for the CO can then implement a `HealNode` function (with Kubernetes medik8s might be an option)

This makes it easier to follow the actions, and improve the reporting to users (Kubernetes events maybe?).

CRDs are Kubernetes specific, so such definitions should go in the kubernetes-csi-addons repository instead. There probably should be a way for the CO to configure the checking interval, maybe even per volume. When using Kubernetes, an annotation on the PVC might be suitable for that (depending on the number of options).